File: packages/browseros-agent/apps/server/src/browser/core/windows.ts
import type { WindowInfo } from '@browseros/cdp-protocol/domains/browser'
import type { CdpConnection } from './connection'

export type { WindowInfo }

export interface SetWindowVisibilityResult {
  window: WindowInfo
  replaced: boolean
  previousWindowId: number
  newWindowId: number
}

const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms))

/** Wraps BrowserOS window CDP commands for browser-core callers and tools. */
export class WindowManager {
  constructor(private readonly cdp: CdpConnection) {}

  async list(): Promise<WindowInfo[]> {
    await this.ensureConnected()
    const result = await this.cdp.Browser.getWindows()
    return result.windows as WindowInfo[]
  }

  async create(opts?: { hidden?: boolean }): Promise<WindowInfo> {
    await this.ensureConnected()
    const result = await this.cdp.Browser.createWindow({
      hidden: opts?.hidden ?? false,
    })
    return result.window as WindowInfo
  }

  async close(windowId: number): Promise<void> {
    await this.ensureConnected()
    await this.cdp.Browser.closeWindow({ windowId })
  }

  async activate(windowId: number): Promise<void> {
    await this.ensureConnected()
    await this.cdp.Browser.activateWindow({ windowId })
  }

  /** Changes visibility and returns the replacement window ID when BrowserOS swaps windows. */
  async setVisibility(
    windowId: number,
    opts: { visible: boolean; activate?: boolean },
  ): Promise<SetWindowVisibilityResult> {
    await this.ensureConnected()
    const result = await this.cdp.Browser.setWindowVisibility({
      windowId,
      visible: opts.visible,
      ...(opts.activate !== undefined && { activate: opts.activate }),
    })
    const window = result.window as WindowInfo
    return {
      window,
      replaced: result.replaced,
      previousWindowId: result.previousWindowId,
      newWindowId: window.windowId,
    }
  }

  private async ensureConnected(): Promise<void> {
    if (this.cdp.isConnected()) return

    const deadline = Date.now() + 5000
    while (!this.cdp.isConnected() && Date.now() < deadline) {
      await delay(50)
    }
    if (!this.cdp.isConnected()) throw new Error('CDP not connected')
  }
}

File: packages/browseros-agent/crates/browseros-core/src/windows.rs
use crate::{CoreError, ProtocolSession, WindowId, connection::CdpConnection, timeouts};
use browseros_cdp::browser;
use serde_json::{Value, json};
use std::sync::Arc;
use tokio::time::sleep;

pub type WindowInfo = browser::WindowInfo;

#[derive(Debug, Clone, PartialEq)]
pub struct SetWindowVisibilityResult {
    pub window: WindowInfo,
    pub replaced: bool,
    pub previous_window_id: WindowId,
    pub new_window_id: WindowId,
}

pub struct WindowManager {
    cdp: Arc<dyn CdpConnection>,
}

impl WindowManager {
    #[must_use]
    pub fn new(cdp: Arc<dyn CdpConnection>) -> Self {
        Self { cdp }
    }

    pub async fn list(&self) -> Result<Vec<WindowInfo>, CoreError> {
        self.ensure_connected().await?;
        let root = ProtocolSession::root(self.cdp.clone());
        let result: browser::GetWindowsResult = root.send("Browser.getWindows", json!({})).await?;
        Ok(result.windows)
    }

    pub async fn create(&self, hidden: bool) -> Result<WindowInfo, CoreError> {
        self.ensure_connected().await?;
        let root = ProtocolSession::root(self.cdp.clone());
        let result: browser::CreateWindowResult = root
            .send("Browser.createWindow", json!({ "hidden": hidden }))
            .await?;
        Ok(result.window)
    }

    pub async fn close(&self, window_id: WindowId) -> Result<(), CoreError> {
        self.ensure_connected().await?;
        let root = ProtocolSession::root(self.cdp.clone());
        let _: Value = root
            .send("Browser.closeWindow", json!({ "windowId": window_id.0 }))
            .await?;
        Ok(())
    }

    pub async fn activate(&self, window_id: WindowId) -> Result<(), CoreError> {
        self.ensure_connected().await?;
        let root = ProtocolSession::root(self.cdp.clone());
        let _: Value = root
            .send("Browser.activateWindow", json!({ "windowId": window_id.0 }))
            .await?;
        Ok(())
    }

    pub async fn set_visibility(
        &self,
        window_id: WindowId,
        visible: bool,
        activate: Option<bool>,
    ) -> Result<SetWindowVisibilityResult, CoreError> {
        self.ensure_connected().await?;
        let root = ProtocolSession::root(self.cdp.clone());
        let result: browser::SetWindowVisibilityResult = root
            .send(
                "Browser.setWindowVisibility",
                json!({ "windowId": window_id.0, "visible": visible, "activate": activate }),
            )
            .await?;
        let new_window_id = WindowId(result.window.window_id);
        Ok(SetWindowVisibilityResult {
            window: result.window,
            replaced: result.replaced,
            previous_window_id: WindowId(result.previous_window_id),
            new_window_id,
        })
    }

    async fn ensure_connected(&self) -> Result<(), CoreError> {
        if self.cdp.is_connected() {
            return Ok(());
        }
        let deadline = tokio::time::Instant::now() + timeouts::WAIT_FOR_CONNECTION_TIMEOUT;
        while !self.cdp.is_connected() && tokio::time::Instant::now() < deadline {
            sleep(timeouts::WAIT_FOR_CONNECTION_POLL).await;
        }
        if self.cdp.is_connected() {
            Ok(())
        } else {
            Err(CoreError::Cdp(browseros_cdp::CdpError::NotConnected))
        }
    }
}
